### PR TITLE
Reduce memory overhead of Tracer#trace

### PR DESCRIPTION
**Pull request by marcotc**
_**TL;DR: 26%-35% memory reduction, 3%-17% speedup**_

This PR aims to reduce the memory footprint of a trace object creation. Execution time improvement is not the goal, but a regression should be avoided.

This work was mostly guided by the results of our benchmark memory report `Microbenchmark benchmark detailed report memory report`. This test provides a very detailed result of all allocation sites exercised during a run.

We use the benchmark `Microbenchmark Datadog::Tracer nested traces` to measure the impact of our changes. This benchmark runs the statement `tracer.trace('my.op'){ ... }` with varying amount of nested spans: `1`, `10` and `100` in the results below represent the number of nested spans created during a test run.

Bellow are the results after the changes have been applied:

#### Before
```
Calculating -------------------------------------
                   1   254.312k memsize (    92.048k retained)
                         2.848k objects (   718.000  retained)
                         4.000  strings (     1.000  retained)
                  10     1.811M memsize (   861.400k retained)
                        14.510k objects (     5.639k retained)
                         3.000  strings (     0.000  retained)
                 100    17.342M memsize (     8.505M retained)
                       131.531k objects (    55.102k retained)
                         3.000  strings (     0.000  retained)

Comparison:
                   1:     254312 allocated
                  10:    1811440 allocated - 7.12x more
                 100:   17341816 allocated - 68.19x more
```
#### After
```
Calculating -------------------------------------
                   1   164.504k memsize (    91.760k retained)
                         2.044k objects (   712.000  retained) 1 - 2.044/2.848
                         1.000  strings (     1.000  retained)
                  10     1.306M memsize (   862.720k retained)
                        11.876k objects (     5.640k retained) 1 - 11.876/14.510
                         1.000  strings (     1.000  retained)
                 100    12.663M memsize (     8.512M retained)
                       110.968k objects (    55.184k retained) 1 - 110.968/131.531
                         1.000  strings (     1.000  retained)

Comparison:
                   1:     164504 allocated
                  10:    1306024 allocated - 7.94x more
                 100:   12663240 allocated - 76.98x more
```
#### Comparison
```
  1: 100% -   164504.0/254312.0   = 0.35314102362452415
 10: 100% -  1306024.0/1811440.0  = 0.27901338161904343
100: 100% - 12663240.0/17341816.0 = 0.269785817125496
```

We have a **reduction of bytes allocated of 26%-35%**, depending on the trace nesting depth.

### Time
Even though time performance was not the goal, it is important to ensure there are not regressions. Here are the timing results:
#### Before
```
Calculating -------------------------------------
                   1     37.323k (±24.6%) i/s -    168.652k in   5.076209s
                  10      9.328k (±37.7%) i/s -     37.098k in   5.844379s
                 100      1.150k (±47.2%) i/s -      3.910k in   5.899024s
Comparison:
                   1:    37322.6 i/s
                  10:     9328.3 i/s - 4.00x  (± 0.00) slower
                 100:     1150.0 i/s - 32.46x  (± 0.00) slower
```
#### After
```
Calculating -------------------------------------
                   1     38.473k (±23.0%) i/s -    175.725k in   5.003995s
                  10      9.606k (±39.2%) i/s -     34.768k in   5.013167s
                 100      1.354k (±39.4%) i/s -      4.484k in   5.054625s
Comparison:
                   1:    38472.9 i/s
                  10:     9606.2 i/s - 4.00x  (± 0.00) slower
                 100:     1354.3 i/s - 28.41x  (± 0.00) slower
```

#### Comparison
```
  1: 38472.9/37322.6 = 1.030820468027415115
 10:  9606.2/9328.3  = 1.029791065896251245
100:  1354.3/1150.0  = 1.17765217391304344
```

We have a minor **speedup of 3%-17%**, depending on the trace nesting depth.

### CODE REVIEW

File: lib/ddtrace/tracer.rb [hunk 1]
```diff
@@ -186,23 +187,24 @@ def guess_context_and_parent(child_of)
     # * +start_time+: when the span actually starts (defaults to \now)
     # * +tags+: extra tags which should be added to the span.
     def start_span(name, options = {})
-      start_time = options.fetch(:start_time, Time.now.utc)
+      start_time = options[:start_time] || Time.now
```
Comment (brettlangdon) [inline lib/ddtrace/tracer.rb:190]: why this change? is `#fetch` slow?

Comment (brettlangdon) [reply lib/ddtrace/tracer.rb:190]: I suggest we add a comment so we don't revert in the future (?)

File: lib/ddtrace/utils.rb [hunk 2]
```diff
@@ -3,6 +3,11 @@
 module Datadog
   # Utils contains low-level utilities, typically to provide pseudo-random trace IDs.
   module Utils
+    # Empty immutable hash to be used.
+    # Useful to avoid allocations when the hash won't be modified:
+    # e.g. methods with empty +options={}+ as default.
+    EMPTY_HASH = {}.freeze
```
Comment (delner) [inline lib/ddtrace/utils.rb:9]: Since this is effectively a constant, should we move this also to `Ext`?

I'm pretty sure we're going to apply what you've done here in many other places as well, and I think it might be better to have those avoid referencing `Utils` as its more of a function library vs `Ext` which is for universal constants.

Minor issue either way.

File: lib/ddtrace/tracer.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/tracer.rb:194]: Added comment!

**Comment by delner** [comment]
@brettlangdon FWIW, if there are specific rules we're undertaking here for performance (frozen strings, empty hash constant, avoiding `fetch` or blocks) then we should probably add something to our development guide that outlines performance-driven suggestions for style.

**Comment by brettlangdon** [comment]
> @brettlangdon FWIW, if there are specific rules we're undertaking here for performance (frozen strings, empty hash constant, avoiding `fetch` or blocks) then we should probably add something to our development guide that outlines performance-driven suggestions for style.
